3rdParty/iresearch/core/formats/columnstore2.cpp

```diff
@@ -1514,41 +1701,62 @@ void reader::prepare_index(
-        data_cipher_->decrypt(offset,
-                              reinterpret_cast<byte_type*>(name->data()),
-                              name->size());
+        data_cipher_->decrypt(
+            offset, reinterpret_cast<byte_type*>(name->data()), name->size());
       }
     }
 
-    auto index = hdr.docs_index
-      ? read_bitmap_index(*index_in)
-      : column_index{};
+    auto index = hdr.docs_index ? read_bitmap_index(*index_in) : column_index{};
 
     if (const size_t idx = static_cast<size_t>(hdr.type);
         IRS_LIKELY(idx < IRESEARCH_COUNTOF(kFactories))) {
-      auto column = kFactories[idx](std::move(name), std::move(payload),
-                                   std::move(hdr), std::move(index),
-                                   *index_in, *data_in_,
-                                   std::move(inflater), data_cipher_.get());
+      // check hotness
+
+      auto column = kFactories[idx](
+          std::move(name), std::move(payload), std::move(hdr),
+                           std::move(index), *index_in, *data_in_,
+                           std::move(inflater), data_cipher_.get());
       assert(column);
 
       if (!sorted_columns.empty() &&
           ::less(column->name(), sorted_columns.back()->name())) {
         throw irs::index_error(irs::string_utils::to_string(
-            "invalid column order in segment '%s'",
-            meta.name.c_str()));
+            "invalid column order in segment '%s'", meta.name.c_str()));
       }
 
       assert(hdr.id < columns.size());
       columns[hdr.id] = column.get();
       sorted_columns.emplace_back(std::move(column));
     } else {
       throw index_error{string_utils::to_string(
-        "Failed to load column id=" IR_SIZE_T_SPECIFIER " , got invalid type=%u",
-        i, static_cast<uint32_t>(hdr.type))};
+          "Failed to load column id=" IR_SIZE_T_SPECIFIER
+          " , got invalid type=%u",
+          i, static_cast<uint32_t>(hdr.type))};
+    }
+  }
+  if (opts.warmup_column) {
+    auto memory_accounting =
+        opts.pinned_memory ? opts.pinned_memory : noop_memory_accounter;
+    for (size_t i = 0; i < sorted_columns.size(); ++i) {
+      auto cb = static_cast<column_base*>(sorted_columns[i].get());
+      if (opts.warmup_column(*cb)) {
+        if (!direct_data_input) {
+          direct_data_input =
+              dir.open(data_filename, IOAdvice::DIRECT_READ);
+          if (!direct_data_input) {
+            IR_FRMT_WARN("Failed to open direct access file, path: %s. Columns buffering stopped.",
+                         data_filename.c_str());
+            break;
+          }
+        }
+        IR_FRMT_TRACE("Making buffered: %d", cb->header().id);
+        cb->make_buffered(*direct_data_input, memory_accounting, 
+                          range(sorted_columns.data() + i + 1,
+                                sorted_columns.size() - i - 1));
+        IR_FRMT_TRACE("Finished buffered: %d", cb->header().id);
```

Comment:
`id` seems to be a `uint64_t`.
```suggestion
        IR_FRMT_TRACE("Finished buffered: %llu", cb->header().id);
```